From ac0c63bf369939322a19c121090c8c94e15b119f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Maxime=20M=C3=A9r=C3=A9?= <maxime.mere@foss.st.com>
Date: Fri, 31 Jan 2025 18:08:55 +0100
Subject: [PATCH 780/841] feat(st-crypto): improve RNG parameter for all
 platform
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The value of RNG_NSCR registe could be improved according to IP version.
Set a better value to improve the IP behavior.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 4b08c936e8bcb0aa243a45730adee3df24d4d7d2 feat(st-crypto): improve RNG parameter for all platform]
Change-Id: I4a4339f684a5ac6929e5ad9296be5749a53bab3d
Signed-off-by: Maxime Méré <maxime.mere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/433667
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/st/crypto/stm32_rng.c | 29 ++++++++++++++++++++++++-----
 1 file changed, 24 insertions(+), 5 deletions(-)

diff --git a/drivers/st/crypto/stm32_rng.c b/drivers/st/crypto/stm32_rng.c
index 407eefb75..dd7bcfcbd 100644
--- a/drivers/st/crypto/stm32_rng.c
+++ b/drivers/st/crypto/stm32_rng.c
@@ -30,6 +30,7 @@
 #define RNG_SR				0x04U
 #define RNG_DR				0x08U
 #if STM32_RNG_VER == 4
+#define RNG_NSCR			0x0CU
 #define RNG_HTCR			0x10U
 #endif
 
@@ -55,18 +56,34 @@
 #define TIMEOUT_US_1MS			1000U
 
 #if STM32_RNG_VER == 4
-#define RNG_NIST_CONFIG1		0xFU
-#define RNG_NIST_CONFIG2		0x0U
-#define RNG_NIST_CONFIG3		0xFU
 #define RNG_NIST_CONFIG(x,y,z)		(((x) << RNG_CR_RNG_CONFIG1_SHIFT) | \
 					 ((y) << RNG_CR_RNG_CONFIG2_SHIFT) | \
 					 ((z) << RNG_CR_RNG_CONFIG3_SHIFT))
 #define RNG_NIST_CONFIG_MASK		GENMASK_32(25, 8)
 
 #if STM32_RNG_VER_MINOR == 2
-#define RNG_HTCFG_CONFIG		0x000072ACU /* Reset value */
-#else
+/* MP13 default values */
+#define RNG_NIST_CONFIG1		0xFU
+#define RNG_NIST_CONFIG2		0x0U
+#define RNG_NIST_CONFIG3		0xDU
+#define RNG_HTCFG_CONFIG		0x0000969DU
+#define RNG_NSCFG_CONFIG		0x0002B5BBU
+#elif STM32_RNG_VER_MINOR == 3
+/* MP25 and MP23 default values */
+#define RNG_NIST_CONFIG1		0x8FU
+#define RNG_NIST_CONFIG2		0x0U
+#define RNG_NIST_CONFIG3		0xEU
+#define RNG_HTCFG_CONFIG		0x00006688U
+#define RNG_NSCFG_CONFIG		0x0002E649U
+#elif STM32_RNG_VER_MINOR == 4
+/* MP21 default values */
+#define RNG_NIST_CONFIG1		0xFU
+#define RNG_NIST_CONFIG2		0x0U
+#define RNG_NIST_CONFIG3		0xFU
 #define RNG_HTCFG_CONFIG		0x0000AAC7U
+#define RNG_NSCFG_CONFIG		0x000001FFU
+#else
+#error "Please define STM32_RNG_VER_MINOR"
 #endif
 #endif
 
@@ -149,6 +166,8 @@ static int stm32_rng_enable(void)
 
 	mmio_write_32(stm32_rng.base + RNG_HTCR, stm32_rng.ht_cfg);
 
+	mmio_write_32(stm32_rng.base + RNG_NSCR, RNG_NSCFG_CONFIG);
+
 	mmio_clrsetbits_32(stm32_rng.base + RNG_CR, RNG_CR_CONDRST, RNG_CR_RNGEN);
 #endif
 	timeout = timeout_init_us(RNG_TIMEOUT_US);
-- 
2.47.2

